File: dev-server.js
#!/usr/bin/env node

/**
 * Servidor de desenvolvimento para o Dashboard Mpinda Evata
 * Serve arquivos estáticos e fornece live reload básico
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');

const PORT = process.env.PORT || 3000;
const HOST = process.env.HOST || 'localhost';

// MIME types para diferentes extensões
const mimeTypes = {
  '.html': 'text/html',
  '.css': 'text/css',
  '.js': 'application/javascript',
  '.json': 'application/json',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon'
};

function serveFile(filePath, res) {
  const ext = path.extname(filePath).toLowerCase();
  const contentType = mimeTypes[ext] || 'application/octet-stream';

  fs.readFile(filePath, (err, content) => {
    if (err) {
      if (err.code === 'ENOENT') {
        // Arquivo não encontrado
        res.writeHead(404, { 'Content-Type': 'text/html; charset=utf-8' });
        res.end(`
          <h1>404 - Arquivo não encontrado</h1>
          <p>O arquivo <code>${filePath}</code> não foi encontrado.</p>
          <p><a href="/">← Voltar ao início</a></p>
        `);
      } else {
        // Erro interno do servidor
        res.writeHead(500, { 'Content-Type': 'text/html; charset=utf-8' });
        res.end(`
          <h1>500 - Erro interno do servidor</h1>
          <p>Erro ao ler o arquivo: ${err.message}</p>
        `);
      }
    } else {
      res.writeHead(200, { 
        'Content-Type': contentType + '; charset=utf-8',
        'Cache-Control': 'no-cache' // Evita cache durante desenvolvimento
      });
      res.end(content);
    }
  });
}

const server = http.createServer((req, res) => {
  // Parse da URL
  const parsedUrl = url.parse(req.url);
  let pathname = parsedUrl.pathname;

  // Remove query string e fragmentos
  pathname = pathname.replace(/\?.*$/, '').replace(/#.*$/, '');

  // Previne directory traversal
  pathname = path.normalize(pathname);
  if (pathname.includes('..')) {
    res.writeHead(400, { 'Content-Type': 'text/html; charset=utf-8' });
    res.end('<h1>400 - Requisição inválida</h1>');
    return;
  }

  // Se é a raiz, serve o index.html
  if (pathname === '/') {
    pathname = '/index.html';
  }

  // Constrói o caminho do arquivo
  const filePath = path.join(__dirname, pathname);

  // Log da requisição
  console.log(`${new Date().toISOString()} - ${req.method} ${pathname}`);

  // Serve o arquivo
  serveFile(filePath, res);
});

server.listen(PORT, HOST, () => {
  console.log('🚀 Dashboard Mpinda Evata - Servidor de Desenvolvimento');
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log(`📍 Servidor rodando em: http://${HOST}:${PORT}`);
  console.log(`📁 Servindo arquivos de: ${__dirname}`);
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log('');
  console.log('💡 Dicas:');
  console.log('   • Acesse http://localhost:3000 para ver o dashboard');
  console.log('   • Pressione Ctrl+C para parar o servidor');
  console.log('   • Arquivos são servidos sem cache para desenvolvimento');
  console.log('');
  
  // Tenta abrir o navegador automaticamente (apenas no Windows)
  if (process.platform === 'win32') {
    const { exec } = require('child_process');
    exec(`start http://${HOST}:${PORT}`, (err) => {
      if (!err) {
        console.log('🌐 Abrindo navegador automaticamente...');
      }
    });
  }
});

// Tratamento de erros
server.on('error', (err) => {
  if (err.code === 'EADDRINUSE') {
    console.error(`❌ Erro: Porta ${PORT} já está em uso.`);
    console.error('   Tente usar uma porta diferente:');
    console.error(`   PORT=3001 node dev-server.js`);
  } else {
    console.error('❌ Erro no servidor:', err.message);
  }
  process.exit(1);
});

// Tratamento de sinais para encerramento gracioso
process.on('SIGINT', () => {
  console.log('\n🛑 Encerrando servidor...');
  server.close(() => {
    console.log('✅ Servidor encerrado com sucesso.');
    process.exit(0);
  });
});

process.on('SIGTERM', () => {
  console.log('\n🛑 Recebido SIGTERM, encerrando servidor...');
  server.close(() => {
    console.log('✅ Servidor encerrado com sucesso.');
    process.exit(0);
  });
});

File: dev-server.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Servidor de desenvolvimento para o Dashboard Mpinda Evata
Serve arquivos estáticos com suporte a CORS e live reload básico
"""

import http.server
import socketserver
import os
import sys
import webbrowser
import threading
import time
from urllib.parse import urlparse

PORT = int(os.environ.get('PORT', 3000))
HOST = os.environ.get('HOST', 'localhost')

class DashboardHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    """Handler customizado para o servidor de desenvolvimento"""
    
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=os.getcwd(), **kwargs)
    
    def end_headers(self):
        # Adiciona headers para desenvolvimento
        self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
        self.send_header('Pragma', 'no-cache')
        self.send_header('Expires', '0')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        super().end_headers()
    
    def log_message(self, format, *args):
        # Log customizado com timestamp
        timestamp = time.strftime('%Y-%m-%d %H:%M:%S')
        sys.stdout.write(f"{timestamp} - {format % args}\n")
    
    def do_GET(self):
        # Se é a raiz, redireciona para index.html
        if self.path == '/':
            self.path = '/index.html'
        
        # Previne directory traversal
        parsed_path = urlparse(self.path)
        if '..' in parsed_path.path:
            self.send_error(400, "Requisição inválida")
            return
        
        return super().do_GET()

def open_browser():
    """Abre o navegador após um pequeno delay"""
    time.sleep(1)
    try:
        webbrowser.open(f'http://{HOST}:{PORT}')
        print(f"🌐 Abrindo navegador automaticamente...")
    except Exception as e:
        print(f"⚠️  Não foi possível abrir o navegador automaticamente: {e}")

def main():
    """Função principal do servidor"""
    try:
        # Configura o servidor
        with socketserver.TCPServer((HOST, PORT), DashboardHTTPRequestHandler) as httpd:
            print('🚀 Dashboard Mpinda Evata - Servidor de Desenvolvimento')
            print('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━')
            print(f'📍 Servidor rodando em: http://{HOST}:{PORT}')
            print(f'📁 Servindo arquivos de: {os.getcwd()}')
            print('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━')
            print('')
            print('💡 Dicas:')
            print('   • Acesse http://localhost:3000 para ver o dashboard')
            print('   • Pressione Ctrl+C para parar o servidor')
            print('   • Arquivos são servidos sem cache para desenvolvimento')
            print('')
            
            # Abre o navegador em uma thread separada
            browser_thread = threading.Thread(target=open_browser)
            browser_thread.daemon = True
            browser_thread.start()
            
            # Inicia o servidor
            httpd.serve_forever()
            
    except OSError as e:
        if e.errno == 98 or 'Address already in use' in str(e):
            print(f"❌ Erro: Porta {PORT} já está em uso.")
            print("   Tente usar uma porta diferente:")
            print(f"   PORT=3001 python dev-server.py")
        else:
            print(f"❌ Erro no servidor: {e}")
        sys.exit(1)
    except KeyboardInterrupt:
        print('\n🛑 Encerrando servidor...')
        print('✅ Servidor encerrado com sucesso.')
        sys.exit(0)

if __name__ == '__main__':
    main()